### PR TITLE
#243: Add type field to LB data files

### LINKED ISSUE #243: Add type field to LB data files
Also, add schema for other JSON files.

### PR DESCRIPTION
**Pull request by marcinwrobel1986**
### THIS PR:
- modified schemaValidator in order to support another schema
- added schema type and support checking
- modified synthetic lbaf data in order to be supported
- fixed all tests
- updated JSON_data_files_validator.py

Closes #243

### CODE REVIEW

**Comment by marcinwrobel1986** [comment]
Hello @nlslatt @lifflander @ppebay ,
please take a look at `src/lbaf/IO/schemaValidator.py` and check if there are some keys, that should be optional
Thanks!

**Review by nlslatt — COMMENTED** [review]
Do we still need tests of the stats schema? I marked some stuff that should be optional, but more discussion is needed to decide on whether other parts can be made optional.

@lifflander We talked early on about potentially allowing only a subset of statistical quantities to be output. Now that we're using json and are planning for a tool that extracts the desired quantities, do we still want to make any of them (e.g., kurtosis) optional in the schema? Do we want to make any of the statistics types (like comm or modeled vs. raw) optional?

File: src/lbaf/Utils/JSON_data_files_validator.py [hunk 1]
```diff
@@ -86,12 +86,17 @@ def __validate_file(file_path):
             except brotli.error:
                 decompressed_dict = json.loads(compr_bytes.decode("utf-8"))
 
+        # Extracting type from JSON data
+        schema_type = decompressed_dict.get("type")
+        if schema_type is None:
+            raise TypeError("JSON data is missing 'type' key")
```
[the remark this answers is not in the document]
Comment (marcinwrobel1986) [reply src/lbaf/Utils/JSON_data_files_validator.py:92]: That is a good question if we are planning to use `JSON_data_files_validator.py` for files other than VT. `LBStatsfile` schema was added to `SchemaValidator`, so at the moment `JSON_data_files_validator.py` checks if type of the data is "LBDatafile" or "LBStatsfile" and then proceed schema checking accordingly.

[the remark this answers is not in the document]
Comment (nlslatt) [reply src/lbaf/Utils/JSON_data_files_validator.py:92]: Let's make it an automatic pass if it doesn't have one of the matching scheme types.

[the remark this answers is not in the document]
Comment (marcinwrobel1986) [reply src/lbaf/Utils/JSON_data_files_validator.py:92]: ok

File: src/lbaf/IO/schemaValidator.py [hunk 2]
```diff
@@ -1,18 +1,28 @@
...
+
+    def _get_valid_schema(self) -> Schema:
         """ Returns representation of a valid schema
         """
-        valid_schema = Schema(
+        allowed_types = ("LBDatafile", "LBStatsfile")
+        valid_schema_data = Schema(
             {
+                'type': And(str, lambda a: a in allowed_types,
+                            error=f"{self.get_error_message(allowed_types)} needs to be chosen"),
```
Comment (ppebay) [inline src/lbaf/IO/schemaValidator.py:25]: needs to ==> must be (everywhere not only here)

File: src/lbaf/IO/schemaValidator.py [hunk 3]
```diff
@@ -68,7 +78,185 @@ def _get_valid_schema() -> Schema:
...
+                ]
+            }
+        )
+
+        if self.schema_type == "LBDatafile":
+            return valid_schema_data
+        elif self.schema_type == "LBStatsfile":
+            return valid_schema_stats
+
+        raise TypeError(f"Not supported schema type: {self.schema_type} were given")
```
Comment (ppebay) [inline src/lbaf/IO/schemaValidator.py:259]: Unsupported

File: src/lbaf/IO/schemaValidator.py [hunk 4]
```diff
@@ -68,7 +78,185 @@ def _get_valid_schema() -> Schema:
...
 
     def is_valid(self, schema_to_validate: dict) -> bool:
         """ Returns True is schema_to_validate is valid with self.valid_schema else False. """
```
Comment (ppebay) [inline src/lbaf/IO/schemaValidator.py:262]: There is a problem in this sentence : I think the first `is` should be `if`

File: src/lbaf/Utils/JSON_data_files_validator.py [hunk 5]
```diff
@@ -86,12 +86,17 @@ def __validate_file(file_path):
             except brotli.error:
                 decompressed_dict = json.loads(compr_bytes.decode("utf-8"))
 
-        # Validate schema
-        if SchemaValidator().is_valid(schema_to_validate=decompressed_dict):
-            print(f"=> Valid JSON schema in {file_path}")
```
Comment (ppebay) [inline src/lbaf/Utils/JSON_data_files_validator.py:91]: No arrow

**Review by ppebay — CHANGES_REQUESTED** [review]

**Review by ppebay — APPROVED** [review]

File: src/lbaf/Utils/JSON_data_files_validator.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ppebay) [reply src/lbaf/Utils/JSON_data_files_validator.py]: Is this resolved?

File: src/lbaf/IO/schemaValidator.py [hunk 6]
```diff
@@ -68,7 +78,185 @@ def _get_valid_schema() -> Schema:
...
-        return valid_schema
+        valid_schema_stats = Schema(
+            {
+                'type': And(str, lambda a: a in allowed_types,
+                            error=f"{self.get_error_message(allowed_types)} needs to be chosen"),
+                'phases': [
+                    {
+                        "id": int,
+                        "migration count": int,
+                        "post-LB": {
```
[the remark this answers is not in the document]
Comment (ppebay) [reply src/lbaf/IO/schemaValidator.py]: Resolved?

File: src/lbaf/Utils/JSON_data_files_validator.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (marcinwrobel1986) [reply src/lbaf/Utils/JSON_data_files_validator.py]: Yes, it's passing as Nicole requested.

**Comment by nlslatt** [comment]
@marcinwrobel1986 Are there any tests of the `LBStatsfile` schema like there are of the `LBDatafile` one?

**Comment by nlslatt** [comment]
@marcinwrobel1986 Please make meaningful commit messages for commit `1d5add0878ca2c93d6ce00a9ce511cbf973efd6c`. `merged develop` does not accurately describe the changes.

**Comment by marcinwrobel1986** [comment]
> @marcinwrobel1986 Please make meaningful commit messages for commit `1d5add0878ca2c93d6ce00a9ce511cbf973efd6c`. `merged develop` does not accurately describe the changes.

Hi @nlslatt sorry, but I missed that the previous message regarding develop merge was sent
Zrzut ekranu z 2022-05-23 22-55-58

**Comment by marcinwrobel1986** [comment]
> @marcinwrobel1986 Are there any tests of the `LBStatsfile` schema like there are of the `LBDatafile` one?

Tests for `JSON_data_files_validator.py` are on the branch `201-..` and will be probably merged tomorrow. I will write some regarding `LBStatsfile`, as well, because there are none at the moment.

**Review by nlslatt — APPROVED** [review]

**Pull request merged**